pkg/store/bucket.go

```diff
@@ -1188,32 +1201,61 @@ func (s *BucketStore) LabelNames(ctx context.Context, req *storepb.LabelNamesReq
 		resHints.AddQueriedBlock(b.meta.ULID)
 
 		indexr := b.indexReader(gctx)
-		extLabels := b.meta.Thanos.Labels
 
 		g.Go(func() error {
 			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label names")
 
-			// Do it via index reader to have pending reader registered correctly.
-			res, err := indexr.block.indexHeaderReader.LabelNames()
-			if err != nil {
-				return errors.Wrap(err, "label names")
-			}
+			var result []string
+			if len(reqSeriesMatchers) == 0 {
+				// Do it via index reader to have pending reader registered correctly.
+				res, err := indexr.block.indexHeaderReader.LabelNames()
+				if err != nil {
+					return errors.Wrapf(err, "label names for block %s", b.meta.ULID)
+				}
 
-			// Add  a set for the external labels as well.
-			// We're not adding them directly to res because there could be duplicates.
-			extRes := make([]string, 0, len(extLabels))
-			for lName := range extLabels {
-				extRes = append(extRes, lName)
-			}
+				sort.Strings(res)
```

Comment:
The result of `indexr.block.indexHeaderReader.LabelNames()` is already sorted https://github.com/thanos-io/thanos/blob/main/pkg/block/indexheader/binary_reader.go#L904.
Do you think we can remove this line?

Reply:
Sounds good, but to make sure we don't break it in the future, I will also update `indexheader.Reader` interface to specify the contract about labels being sorted.